### PR TITLE
Move from ClickOnce to Squirrel.Windows

### LINKED ISSUE #125: move installer over to GitHub releases
We don't actually need any external infrastructure for the installer. Mostly because I'm not sure when the last time was that we published there, and if the original link even works.

Here's my current thinking - use `Squirrel.Windows` and GitHub releases.

This [guide](https://github.com/Squirrel/Squirrel.Windows/blob/master/docs/using/github.md) is what I'd start working from, but the other tasks include:

 - update the website to refer to `https://github.com/Code52/carnac/releases/latest`
 - ???

### LINKED ISSUE #135: checksum the file and add it to the chocolatey installer
For an example of this: https://github.com/shiftkey/chocolatey-beyondcompare/blob/76ad6c026584b88e852cec030119bfb3b89da1e6/src/tools/chocolateyInstall.ps1

### LINKED ISSUE #140: Should we move the build script to Cake?
It's got addins for nuget packaging, chocolatey and squirrel.windows

### PR DESCRIPTION
**Pull request by alfhenrik**
- [x] Introducing Squirrel.Windows
  - [x] Test checking for updates locally in debug mode
  - [ ] Test checking for updates on :octocat: in release mode
- [x] Introducing Cake
  - [x] Build solution
  - [x] Run unit tests
  - [x] Generate Zip file
    - [x] Zip file needs reworking
  - [x] Generate Chocolatey package
  - [x] Generate NuGet package ready to be squirrelified
  - [x] Releasify the NuGet package
  - [x] Publish Squirrel release files to :octocat: 
- [x] Chocolatey
  - [x] Use `ChocolateyPack` to generate `chocolatey` package
  - [x] Add checksum to install script
- [x] Remove old, redundant build script
- [x] Update README, probably...
- [x] Bump version??
- [x] Update authors and components

Closes #125 
Closes #135 
Closes #140

### CODE REVIEW

**Comment by alfhenrik** [comment]
I haven't forgotten about this...Just been a bit busy last few days 🙂

**Comment by shiftkey** [comment]
@alfhenrik no worries. Life Happens™

**Comment by alfhenrik** [comment]
Apart from testing the Squirrel update logic now that the build generates the correct artifacts, this is probably ready for a first-pass review 🤔

**Comment by alfhenrik** [comment]
sad 🐼 ...to build the Squirrel update package it looks like it needs the package of the previous version otherwise it won't update the app...i.e. I build version `0.0.1` and get that installed, I run a new build for version `0.0.2` (which cleans the deploy folder where the previous package was), I run the app again and it won't update to version `0.0.2`. If I don't clean the previous artifacts when building `0.0.2` and I run the app it will successfully update to `0.0.2`...

**Comment by paulcbetts** [comment]
You need to have the current version of the app when Releasing in order to build the new version of the app because of Delta updates. There's a tool in Squirrel called SyncReleases that you can use in CI to pull it down, so you don't have to keep a releases folder in perpetuity.

**Comment by alfhenrik** [comment]
OK, got the `SyncReleases` logic in, now I'm hitting the :octocat: API rate limit for unauthenticated calls (makes sense). 

If I add our auth token as a secure environment variable in `appveyor.yml` and pass that into / use that in the build script am I opening a security hole where the token somehow could leak and be exposed to someone that shouldn't see it...? I've done some googling but haven't been able to find a definitive answer...

**Comment by shiftkey** [comment]
> If I add our auth token as a secure environment variable in appveyor.yml and pass that into / use that in the build script am I opening a security hole where the token somehow could leak and be exposed to someone that shouldn't see it...?

I think you're fine to do this as long as it's encrypted, as per [this link](https://www.appveyor.com/docs/build-configuration/#secure-variables):

> However, secure variables are not decoded during Pull Request builds which prevents someone from submitting PR with malicious build script displaying those variables.

**Comment by alfhenrik** [comment]
Yea, I was going to use the one we already have in the config file.

**Comment by alfhenrik** [comment]
OK, I think the only thing left for me to test at this moment is the app update from :octocat:.

Although, 1 thing I'd like to discuss is the *where should we do the app update check*. 

Currently I do it at startup, but this might not be the best place...and also, do we want to show somewhere in the UI that the app has been updated (so the user knows they need to restart to get the latest...)? E.g. a tooltip once the update manager has successfully pulled down and applied the update.

**Comment by shiftkey** [comment]
> Currently I do it at startup, but this might not be the best place...

Maybe not best to do it immediately at setup, but a few minutes after startup (using a background timer or something?)

> do we want to show somewhere in the UI that the app has been updated (so the user knows they need to restart to get the latest...)?

Let's punt this to another time. If the update is applied on next startup then that should be fine.

**Comment by alfhenrik** [comment]
So something like this? (Excuse my Rx n00bness if this is not correct...)

```
            Observable
                .Timer(TimeSpan.FromMinutes(5))
                .Subscribe(async x =>
                {

                    try
                    {
#if DEBUG
                        using (var mgr = new UpdateManager(carnacUpdateUrl))
                        {
                            await mgr.UpdateApp();
                        }
#else
                using (var mgr = UpdateManager.GitHubUpdateManager(carnacUpdateUrl))
                {
                    await mgr.Result.UpdateApp();
                }
#endif
                    }
                    catch
                    {
                        // Do something useful with the exception
                    }
                });

```

**Comment by alfhenrik** [comment]
Just realised that the `zip` file used by `choco install` won't install the app so it can be updated via `Squirrel` as it right now will just extract the carnac binaries. 

Will Fix!

**Comment by alfhenrik** [comment]
All fixed, and ready for review now 👍

File: build.cake [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	{
+		var syncReleasesDir = toolsDir + Directory("squirrel.windows/tools");
+
+		EnsureDirectoryExists(deployDir);
+		EnsureDirectoryExists(squirrelDeployDir);
+
+		// Create nuget package
+		var releaseFiles = new HashSet<string>(
+				GetFiles(buildDir.Path + "/**/*.*").Select(f => f.FullPath)
+					.Concat(GetFiles((toolsDir + Directory("DeltaCompressionDotNet/lib/net45")).Path + "/*.dll").Select(f => f.FullPath)
```
Comment (shiftkey) [inline build.cake]: Can we make this a bit more readable by tidying up the indenting? e.g

```
var allFiles = GetFiles(buildDir.Path + "/**/*.*").Select(f => f.FullPath);
var deltaCompression = GetFiles((toolsDir + Directory("DeltaCompressionDotNet/lib/net45")).Path + "/*.dll").Select(f => f.FullPath);
...

var releaseFiles = new HashSet<string>(
   allFiles
     .Concat(deltaCompression)
     .Concat(...)
```

File: build.cake [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+		};
+		NuGetPack("./src/Carnac/Carnac.nuspec", nuGetPackSettings);
+		
+		// Sync latest release to build new package
+		var squirrelSyncReleasesExe = syncReleasesDir + File("SyncReleases.exe");
+		StartProcess(squirrelSyncReleasesExe, new ProcessSettings { Arguments = "--url " + githubRepoUrl + " --releaseDir " + squirrelReleaseDir.Path + (!string.IsNullOrEmpty(githubAuthToken) ? " --token " + githubAuthToken : "") });
+
+		// Create new squirrel package
+		Squirrel(
+			squirrelDeployDir + File("carnac." + version + ".nupkg"),
```
Comment (shiftkey) [inline build.cake]: Could we do some string interpolation here?

File: build.cake [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+		var chocoSourceDir = Directory("./src/Chocolatey");
+		var chocoToolsDir = chocoSourceDir + Directory("tools");
+		var chocoInstallFile = chocoToolsDir + File("chocolateyinstall.ps1");
+		var chocoSpecPath = chocoSourceDir + File("carnac.nuspec");
+		var chocoDeployDir = deployDir + Directory("Chocolatey");
+		
+		EnsureDirectoryExists(deployDir);
+		EnsureDirectoryExists(chocoDeployDir);
+
+		var url = /*configuration == "Debug"
```
Comment (shiftkey) [inline build.cake]: Is it worth cleaning up this commented out code?

File: build.cake [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+			}
+		);
+	});
+
+Task("Package-Zip")
+	.IsDependentOn("Package-Squirrel")
+	.Does(() =>
+	{
+		var gitHubDeployDir = deployDir + Directory("GitHub");
+		var zipFile = gitHubDeployDir + File("carnac." + version + ".zip");
```
Comment (shiftkey) [inline build.cake]: String interpolation?

File: src/Carnac/App.xaml.cs [hunk 5]
```diff
@@ -44,6 +56,30 @@ protected override void OnStartup(StartupEventArgs e)
             carnac = new KeysController(keyShowViewModel.Messages, messageProvider, new ConcurrencyService(), settingsProvider);
             carnac.Start();
 
+            Observable
+                .Timer(TimeSpan.FromMinutes(5))
+                .Subscribe(async x =>
+                {
+                    try
+                    {
+#if DEBUG
```
Comment (shiftkey) [inline src/Carnac/App.xaml.cs]: What benefits to we get by testing the updating for debug builds? If it's fiddly and more hassle than it's worth I'd be tempted to just drop this and only enable it for release builds.

File: src/Carnac/UI/PreferencesViewModel.cs [hunk 6]
```diff
@@ -76,14 +76,16 @@ public string Version
                                                          "Paul Jenkins",
                                                          "Dmitry Pursanov",
                                                          "Chris Sainty",
-                                                         "Andrew Tobin"
+                                                         "Andrew Tobin",
+                                                         "Henrik Andersson"
```
Comment (shiftkey) [inline src/Carnac/UI/PreferencesViewModel.cs]: :tada:

File: build.cake (hunk 2 above)
Comment (alfhenrik) [inline build.cake]: Why, yes we can...just need to use the `-Experimental` flag on `Cake` apparently 😄

File: src/Carnac/App.xaml.cs (hunk 5 above)
Comment (alfhenrik) [inline src/Carnac/App.xaml.cs]: Not much, I think it was just more for my benefit while testing. I'll just enable it for release builds

File: src/Carnac/UI/PreferencesViewModel.cs (hunk 6 above)
Comment (alfhenrik) [inline src/Carnac/UI/PreferencesViewModel.cs]: I was hesitating adding **myself** but thought, what the heck...so I did 😀

**Comment by alfhenrik** [comment]
> let me know if there's specific areas or things you'd like me to review.
I think I just wanted a sanity check really 😃

**Comment by alfhenrik** [comment]
I can't think of anything left to do on this, so should we merge it?

**Comment by shiftkey** [comment]
> I can't think of anything left to do on this, so should we merge it?

Yeah, let's give it a shot :shipit:

**Comment by shiftkey** [comment]
@alfhenrik let me know if there's anything else to merge

**Comment by alfhenrik** [comment]
@shiftkey I'm pretty sure everything I wanted to get in is in, at least I can't think of anything else right now 😄

**Pull request merged**